presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolsMapper.java

```diff
@@ -0,0 +1,121 @@
+import com.facebook.presto.sql.planner.plan.AggregationNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.PlanNodeId;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.ExpressionRewriter;
+import com.facebook.presto.sql.tree.ExpressionTreeRewriter;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.facebook.presto.sql.tree.SymbolReference;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+public class SymbolsMapper
+{
+    private final Map<Symbol, Symbol> mapping;
+
+    public SymbolsMapper(Map<Symbol, Symbol> mapping)
+    {
+        this.mapping = requireNonNull(mapping, "mapping is null");
+    }
+
+    public Symbol map(Symbol symbol)
+    {
+        Symbol canonical = symbol;
+        while (mapping.containsKey(canonical)) {
+            canonical = mapping.get(canonical);
+        }
+        return canonical;
+    }
+
+    public Optional<Symbol> map(Optional<Symbol> symbol)
+    {
+        if (symbol.isPresent()) {
+            return Optional.of(map(symbol.get()));
+        }
+        return Optional.empty();
+    }
+
+    public Expression map(Expression value)
+    {
+        return ExpressionTreeRewriter.rewriteWith(new ExpressionRewriter<Void>()
+        {
+            @Override
+            public Expression rewriteSymbolReference(SymbolReference node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+            {
+                Symbol canonical = map(Symbol.from(node));
+                return canonical.toSymbolReference();
+            }
+        }, value);
+    }
+
+    public AggregationNode map(AggregationNode node, PlanNode source, PlanNodeId nodeId)
+    {
+        ImmutableMap.Builder<Symbol, Signature> functionInfos = ImmutableMap.builder();
+        ImmutableMap.Builder<Symbol, FunctionCall> functionCalls = ImmutableMap.builder();
+        ImmutableMap.Builder<Symbol, Symbol> masks = ImmutableMap.builder();
+        for (Map.Entry<Symbol, FunctionCall> entry : node.getAggregations().entrySet()) {
+            Symbol symbol = entry.getKey();
+            Symbol canonical = map(symbol);
+            FunctionCall canonicalCall = (FunctionCall) map(entry.getValue());
+            functionCalls.put(canonical, canonicalCall);
+            functionInfos.put(canonical, node.getFunctions().get(symbol));
+        }
+        for (Map.Entry<Symbol, Symbol> entry : node.getMasks().entrySet()) {
+            masks.put(map(entry.getKey()), map(entry.getValue()));
+        }
+
+        List<List<Symbol>> groupingSets = node.getGroupingSets().stream()
+                .map(this::mapAndDistinct)
```

Comment:
copied from `UnaliasSymbolReferences`. Changing this here will make it harder to refactor UnaliasSymbolReferences in the future.